activesupport/lib/active_support/core_ext/date/calculations.rb

```diff
@@ -133,14 +133,26 @@ def change(options)
     )
   end
 
-  # Allow Date to be compared with Time by converting to DateTime and relying on the <=> from there.
+  # Allow Date to be compared with Time by converting to either a Time object or a DateTime object and relying on
+  # the <=> from there. Using DateTime allows us to set the TZ the date conversion happens in. We want this
+  # behaviour when the user has specifically set a Time.zone. If no Time.zone has been set then we fall back to
+  # the default behaviour of to_time which as of ruby 2.4 respects the processes timezone.
   def compare_with_coercion(other)
     if other.is_a?(Time)
-      to_datetime <=> other
+      convert_for_comparison_with_time <=> other
     else
       compare_without_coercion(other)
     end
   end
   alias_method :compare_without_coercion, :<=>
   alias_method :<=>, :compare_with_coercion
+
+  private
+    def convert_for_comparison_with_time
+      if Time.zone
+        DateTime.new(year, month, day, 0, 0, 0, Time.zone.now.formatted_offset)
```

Comment:
This part is definitely wrong: the offset at `now` and the offset at a given date may be wildly different, between DST and actual TZ changes.